Introduce PseudoElementIdentifier

nt1m opened the pull request:
#### 412734d26e78c3750bbd702e323d11d22c36aaf0
<pre>
Introduce PseudoElementIdentifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=268412">https://bugs.webkit.org/show_bug.cgi?id=268412</a>
<a href="https://rdar.apple.com/121959648">rdar://121959648</a>

Reviewed by Darin Adler.

We want to separate this from PseudoElementRequest so we can re-use it in other places.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::getUncachedPseudoStyle const):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRules):
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::addAuthorKeyframeRules):
* Source/WebCore/style/PseudoElementIdentifier.h: Copied from Source/WebCore/style/PseudoElementRequest.h.
(WebCore::Style::PseudoElementIdentifier::PseudoElementIdentifier):
* Source/WebCore/style/PseudoElementRequest.h:
(WebCore::Style::PseudoElementRequest::PseudoElementRequest):
(WebCore::Style::PseudoElementRequest::identifier const):
(WebCore::Style::PseudoElementRequest::pseudoId const):
(WebCore::Style::PseudoElementRequest::pseudoElementNameArgument const):
(WebCore::Style::PseudoElementRequest::scrollbarState const):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):

Canonical link: <a href="https://commits.webkit.org/273809@main">https://commits.webkit.org/273809@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97961846b7b474e6709f8ce7fa79c5a497def3c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36736 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15672 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38973 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39379 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37981 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18156 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12783 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37298 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13203 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11552 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32733 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40628 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33268 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33090 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37488 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9660 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/35615 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4759 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12735 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/style/PseudoElementIdentifier.h:
```diff
@@ -0,0 +1,53 @@
+namespace WebCore::Style {
+
+struct PseudoElementIdentifier {
+public:
```
We don't need `public` if we want this to be a `struct` (but do we?)

nt1m replied:
ah right I copy pasta this

nt1m replied:
Fixed

darinadler commented on Source/WebCore/style/PseudoElementIdentifier.h:
```diff
@@ -0,0 +1,52 @@
+
+namespace WebCore::Style {
+
+struct PseudoElementIdentifier {
```
Is this `struct` really better with the constructors? Sometimes I like to just let a `struct` be a `struct`.

nt1m replied:
The constructors are mostly there for the ASSERT().

darinadler replied:
That doesn't create an invariant; I am not sure it's worth it.

nt1m replied:
Removed

darinadler commented on Source/WebCore/style/PseudoElementRequest.h:
```diff
@@ -33,23 +34,23 @@ namespace WebCore::Style {
-    // highlight name for ::highlight or view transition name for view transition pseudo elements.
-    AtomString pseudoElementNameArgument;
+private:
+    PseudoElementIdentifier m_identifier;
```
Should this only be an implementation detail? Maybe we want to expose it as a constructor argument or expose a getter?

nt1m replied:
Yeah that seems like a good idea.

nt1m committed:
Introduce PseudoElementIdentifier
https://bugs.webkit.org/show_bug.cgi?id=268412
rdar://121959648

Reviewed by Darin Adler.

We want to separate this from PseudoElementRequest so we can re-use it in other places.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::getUncachedPseudoStyle const):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRules):
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::addAuthorKeyframeRules):
* Source/WebCore/style/PseudoElementIdentifier.h: Copied from Source/WebCore/style/PseudoElementRequest.h.
(WebCore::Style::PseudoElementIdentifier::PseudoElementIdentifier):
* Source/WebCore/style/PseudoElementRequest.h:
(WebCore::Style::PseudoElementRequest::PseudoElementRequest):
(WebCore::Style::PseudoElementRequest::identifier const):
(WebCore::Style::PseudoElementRequest::pseudoId const):
(WebCore::Style::PseudoElementRequest::pseudoElementNameArgument const):
(WebCore::Style::PseudoElementRequest::scrollbarState const):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):

Canonical link: https://commits.webkit.org/273809@main
